File: pset6/cash/cash.py
# cash.py
#
# Keith R Bergerstock
# date: 2018.04.16


import sys


def inputString(prompt):
    """ returns a string from stdin """
    sys.stdout.write(prompt)
    sys.stdout.flush()
    tstr = sys.stdin.readline()
    return tstr.rstrip()


def inputInt(prompt):
    v = inputString(prompt)
    if v.isdigit():
        return int(v)
    else:
        return False


def inputFloat(prompt):
    v = inputString(prompt)
    for c in v:
        if not(c.isdecimal() or c == '.'):
            return False
    return float(v)


def getCoins(amount):
    """ returns the number of coins in the given amout """
    d = [0.25, 0.10, 0.05, 0.01]
    coins = 0.0
    for t in d:
        c = amount // t     # get the integer portion of the denomination
        coins += c          # sum the denominations
        amount -= c * t     # subtract from amount the current denimination value
        amount += 0.000001
    return int(coins)


def main():
    change = -1

    while(change < 0):
        r = inputFloat("How Much change is owed? ")
        if not r:
            change = -1
        elif r > 0:
            change = r
        else:
            change = 0

    if change > 0:
        print("{:d}".format(getCoins(change)))


if __name__ == "__main__":
    main()

File: pset1/cash/cash.c
/*
    Keith R Bergerstock
    krbergerstock@e4kountdown.com
    CS50 - greedy.c

*/


#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>

// retuns as an integer the number of coins required to return the change
int get_coins(float amount)
{
    const float d[4] = { 0.25, 0.10, 0.05, 0.01 };
    int coins = 0;
    float t = 0;
    int i;
    for (i = 0; i < 4; i++)
    {
        t = floor(amount / d[i]);   // obtain the interger portion of denomination
        coins += (int)t;            // sum the denominations
        amount -= t * d[i];         // subtract denomination 8 qty from the amount
        amount += 0.00001;          // add in a correction factor
    }

    return coins;
}

int main(void)
{
    int coins = 0;
    float change = 0.0;
    const char buffer[81];
    do
    {
        printf("How much change is owed? ");
        fgets((char *)buffer, 80,  stdin);
        change = atof(buffer);
    }
    while (!isdigit(*buffer) || change < 0.0);

    if (change > 0.0)
    {
        coins = get_coins(change);
        printf("%d\n", coins);
    }

    return 0;
}